core/pkg/artifacts/downloader.go

```diff
@@ -100,50 +223,68 @@ func (ad *ArtifactDownloader) downloadFiles(artifactID string, manifest Manifest
 	manifestEntries := manifest.Contents
 	numInProgress, numDone := 0, 0
 	nameToScheduledTime := map[string]time.Time{}
-	taskResultsChan := make(chan TaskResult)
+	taskResultsChan := make(chan TaskResult, MAX_BACKLOG)
 	manifestEntriesBatch := make([]ManifestEntry, 0, batchSize)
+	manifestEntriesCopy := map[string]ManifestEntry{}
+	maps.Copy(manifestEntriesCopy, manifestEntries)
+
+	artifactFieldNames, err := GetGraphQLFields(ad.Ctx, ad.GraphqlClient, "Artifact")
+	if err != nil {
+		return err
+	}
+	canFetchFilesByManifestEntry := slices.Contains(artifactFieldNames, "filesByManifestEntries")
 
 	for numDone < len(manifestEntries) {
 		var cursor *string
 		hasNextPage := true
 		for hasNextPage {
 			// Prepare a batch
-			now := time.Now()
 			manifestEntriesBatch = manifestEntriesBatch[:0]
-			response, err := gql.ArtifactFileURLs(
-				ad.Ctx,
-				ad.GraphqlClient,
-				artifactID,
-				cursor,
-				&batchSize,
-			)
-			if err != nil {
-				return err
-			}
-			hasNextPage = response.Artifact.Files.PageInfo.HasNextPage
-			cursor = response.Artifact.Files.PageInfo.EndCursor
-			for _, edge := range response.GetArtifact().GetFiles().Edges {
-				filePath := edge.GetNode().Name
-				entry, err := manifest.GetManifestEntryFromArtifactFilePath(filePath)
+
+			if canFetchFilesByManifestEntry {
```

Comment:
nit -> Can we put the logic from these conditions into helpers? 

It also seems like here, we're checking for the query's existence for every batch..? Maybe we should do that before the `for` loop on line 237. `for numDone < len(manifestEntries) {`

Reply:
yeah definitely can make the logic more modularized in the conditions!

the gql query is made before the for loop, but which query we use determines how we construct each batch so I think we'll still need to check this each loop. let me know if you had something else in mind though!

Reply:
I guess I just meant that once we know that the server doesn't support the new query, every batch can be made with the assumption that the server is old. So I'm not sure why we need the conditional inside the loop that creates the batches. Or am i missing something?

Reply:
[resolved offline]